### PR TITLE
bug 1197444 - Load just-in-time assets in development environment

### PR DESCRIPTION
**Pull request by jwhitlock**
Add jingo helpers to get paths to bundled CSS and JS assets, and add to the in-page global mdn.assets. Use these paths rather than hard-coded values when dynamically loading JS and CSS.

In a development environment, style-prism.js is split into 5 files, and the first one has to be fully loaded before the rest will load correctly.

### CODE REVIEW

File: kuma/static/js/wiki.js [hunk 1]
```diff
@@ -306,19 +316,24 @@
...
+        var $compatCSS, $compatJS;
         if(!$compatTables.length) return;
 
         // don't run if waffle not active
         if(!win.waffle || !win.waffle.flag_is_active('compat_api')) return;
 
+        if(!win.mdn || !mdn.assets) return;
+        /* This will break if these assets are split into multiple files */
+        $compatCSS = mdn.assets.css['wiki-compat-tables'][0];
+        $compatJS = mdn.assets.js['wiki-compat-tables'][0];
```
Comment (openjck) [inline kuma/static/js/wiki.js]: In JavaScript, the `$` character conventionally used to indicate that the variable represents a JQuery object.

Comment (jwhitlock) [inline kuma/static/js/wiki.js]: OK, I'll remove it for these items

**Comment by jwhitlock** [comment]
Force pushed to change JS variables to remove ``$`` (i.e. ``$compatCSS`` to ``compatCSS``) when the value is not a jQuery object.

**Comment by openjck** [comment]
> Force pushed to change JS variables to remove $ (i.e. $compatCSS to compatCSS) when the value is not a jQuery object.

Awesome! Looks good. I'll test functionality and dive deeper into the code tomorrow.

This is great, by the way. I know it'll be a big improvement. :smile:

**Comment by openjck** [comment]
This works beautifully. It might take me a short while to review all the code. Mostly I just want to be sure I understand everything before the migration to django-pipeline.

File: templates/includes/config.html [hunk 2]
```diff
@@ -30,6 +30,16 @@
...
+            },
+            // Just-in-time asset paths
+            assets: {
+                css: {
+                    "wiki-compat-tables": {{ css_bundle_urls("wiki-compat-tables") }}
+                },
+                js: {
+                    "syntax-prism": {{ js_bundle_urls("syntax-prism") }},
+                    "wiki-compat-tables": {{ js_bundle_urls("wiki-compat-tables") }}
+                }
```
Comment (openjck) [inline templates/includes/config.html]: Just a small nitpick. We don't have strict rules about JavaScript coding conventions, but tend to use single quotes around strings.

File: kuma/wiki/helpers.py [hunk 3]
```diff
@@ -232,3 +233,24 @@ def wiki_url(path):
     Look for a wiki page in the current locale, or default to given path
     """
     return reverse('wiki.document', args=[path])
+
+
+def url_list(urls):
+    js_list = '[' + ', '.join('"%s"' % url for url in urls) + ']'
+    return jinja2.Markup(js_list)
```
Comment (openjck) [inline kuma/wiki/helpers.py]: Would `asset_list` be a better name? Assuming I understand correctly that it could represent a list of CSS *or* a list of JS assets.

Comment (openjck) [inline kuma/wiki/helpers.py]: Ah. Makes sense now. This is building a JavaScript array for use in `config.html`. That makes sense. I think the names tripped me up a bit, partly because the term array is more conventional in JS-land. Maybe `def javascript_asset_array` and `js_array` or similar?

File: kuma/wiki/helpers.py [hunk 4]
```diff
@@ -232,3 +233,24 @@ def wiki_url(path):
...
+
+
+@register.function
+def js_bundle_urls(bundle, debug=None):
+    """
+    Return a list of the URLS for a minified and bundled JS file
```
Comment (openjck) [inline kuma/wiki/helpers.py]: Maybe we can use the term *JavaScript array* instead of *list* here as well. Also, my initial reading of the comment led me to think that only minified URLs would be returned.

File: kuma/static/js/wiki.js [hunk 5]
```diff
@@ -306,19 +316,24 @@
...
+        var compatCSS, compatJS;
         if(!$compatTables.length) return;
 
         // don't run if waffle not active
         if(!win.waffle || !win.waffle.flag_is_active('compat_api')) return;
 
+        if(!win.mdn || !mdn.assets) return;
+        /* This will break if these assets are split into multiple files */
+        compatCSS = mdn.assets.css['wiki-compat-tables'][0];
+        compatJS = mdn.assets.js['wiki-compat-tables'][0];
```
Comment (openjck) [inline kuma/static/js/wiki.js]: Awesome! Thanks for being clear about this. We could even make the `forEach` loop (above) more general to handle both the `syntax-prism` bundle and the `wiki-compat-tables` bundle, and any other variable-length bundles we want to dynamically inject in the future.

File: kuma/static/js/wiki.js [hunk 6]
```diff
@@ -191,11 +191,21 @@
         Syntax highlighting scripts
     */
     if($('article pre').length && ('querySelectorAll' in doc)) (function() {
-        var syntaxScript = doc.createElement('script');
-        syntaxScript.setAttribute('data-manual', '');
-        syntaxScript.async = 'true';
-        syntaxScript.src = mdn.staticPath + 'js/syntax-prism-min.js?build=' + mdn.build;
-        doc.body.appendChild(syntaxScript);
+        var bundle = 'syntax-prism';
+        if (window.mdn && mdn.assets.js && mdn.assets.js.hasOwnProperty(bundle)) {
```
Comment (openjck) [inline kuma/static/js/wiki.js]: We assume the existence of `mdn` elsewhere, so we can probably do it here too by omitting the `window.mdn` check.

Comment (openjck) [inline kuma/static/js/wiki.js]: The hasOwnProperty check should go inside the `forEach` loop.

File: kuma/static/js/wiki.js [hunk 7]
```diff
@@ -191,11 +191,21 @@
...
+            mdn.assets.js[bundle].forEach(function(bundle_url, index, array) {
+                var syntaxScript = doc.createElement('script');
+                syntaxScript.setAttribute('data-manual', '');
+                syntaxScript.src = bundle_url;
+                if (index === 0 && array.length > 1) {
+                    /* In development, slow load first script */
```
Comment (openjck) [inline kuma/static/js/wiki.js]: Thanks for adding this comment. Would it be possible to extend it further to explain why?

File: templates/includes/config.html (hunk 2 above)
Comment (jwhitlock) [inline templates/includes/config.html]: OK, will change to single quotes

File: kuma/wiki/helpers.py (hunk 3 above)
Comment (jwhitlock) [inline kuma/wiki/helpers.py]: how about ``asset_url_array``?

File: kuma/wiki/helpers.py (hunk 4 above)
Comment (jwhitlock) [inline kuma/wiki/helpers.py]: I'm expanding the doc strings with some examples to make it clearer, and changing the names to ``js_url_array`` and ``css_url_array``.

File: kuma/static/js/wiki.js (hunk 5 above)
Comment (jwhitlock) [inline kuma/static/js/wiki.js]: I thought about that, but I can't see a clean way to do the chaining (load CSS, then load JS, then run some JS) in the case of multiple files in a bundle.  The specific code seemed cleaner to me.

File: kuma/static/js/wiki.js (hunk 6 above)
Comment (jwhitlock) [inline kuma/static/js/wiki.js]: I'm iterating over the list returned from ``mdn.assets.js['syntax-prism']``, after ``mdn.assets.js.hasOwnProperty('syntax-prism')`` checks that the key exists.  That check doesn't make sense inside the loop.

File: kuma/wiki/helpers.py (hunk 3 above)
Comment (openjck) [inline kuma/wiki/helpers.py]: Makes sense to me :smile:

File: kuma/static/js/wiki.js (hunk 7 above)
Comment (jwhitlock) [inline kuma/static/js/wiki.js]: Yes, I'll add more to the comment.  In production, ``syntax-prism-min.js`` contains all five JS files together, evaluated in order.  In development, they are loaded as five separate files.  The later four refer to ``Prism``, which is defined in the first JS file.  We need to slow down an let the first JS file load completely before loading the follow-on files, or get errors about Prism being undefined.

Comment (jwhitlock) [inline kuma/static/js/wiki.js]: On further testing, ``async = 'false'`` is a truthy value too, so this code does nothing.  Unless we use something stronger (like [require.js](http://www.requirejs.org)), we can't guarentee load order.  Luckily, a refresh usually fixes it.  Changing the comment to an apology.  See [SO 9711160](http://stackoverflow.com/questions/9711160/jquery-load-scripts-in-order) for more ideas.

**Comment by jwhitlock** [comment]
Force pushed:

* Change helpers ``css_bundle_urls`` to ``css_url_array``, ``js_bundle_urls`` to ``js_url_array``, to match that they generate JavaScript arrays.
* Better docstrings for helpers, with output examples
* Fix helper tests in TravisCI
* Simplify belt-and-braces checks for loading syntax-prism
* Add note that syntax-prism loading is unreliable in dev mode, will require a lot of work to fix
* Use single-quotes in JS code, matching current convention

**Comment by darkwing** [comment]
Not to be a nit-picker but in JS we should camelCase instead of using "_" and "-" when naming things.

**Comment by jwhitlock** [comment]
@darkwing - helper.py is Python, ``css_url_array`` and ``js_url_array`` are Jinja2 context functions

**Comment by darkwing** [comment]
I was referring to `bundle_url`

**Comment by jwhitlock** [comment]
Force push: ``bundle_url`` -> ``url`` in /kuma/static/js/wiki.js

File: kuma/static/js/wiki.js [hunk 8]
```diff
@@ -191,11 +191,26 @@
...
+                /*
+                   Note: In development, five scripts are loaded, and the
+                   later scripts use Prism, which is declared in the first
+                   script.  This means syntax highlighting often doesn't work
+                   on the first page load. Refresh, and syntax highlighing
+                   should work.
+
+                   To fix this, we'd have to use something like require.js or
+                   in-client merging of JS files.  Maybe django-pipeline will
+                   help as well.
```
Comment (openjck) [inline kuma/static/js/wiki.js]: `async` is a boolean attribute, so any value (even `async="false"`) will enable asynchronous loading. If the `async` attribute is omitted when injecting multiple scripts, the scripts will execute sequentially.

**Comment by openjck** [comment]
I'm afraid this branch needs a rebase. It also looks like tests are failing.

Thanks for your patience as we go through this. It works beautifully and is going to be a huge improvement over what we have now.

File: kuma/static/js/wiki.js (hunk 8 above)
Comment (jwhitlock) [inline kuma/static/js/wiki.js]: yes, I tried omitting async as well.  It's up to the browser though, and my test browser continued to load the async scripts first, because they were much smaller and downloaded faster.  I decided to leave out async / no async, to prevent people from thinking it was a solved problem.

**Comment by jwhitlock** [comment]
Rebased, tests pass

**Comment by openjck** [comment]
I was hoping to give this one last review before PTO tomorrow but got caught up in Firefox Growth work. If someone wants to grab in the meantime please feel free, or I'll be able to give it more attention next week.

**Comment by openjck** [comment]
Anything more to say on the JavaScript front, @darkwing? Otherwise this is r+ from me.

**Comment by darkwing** [comment]
When I set `DEV` to `False`, I get a `Prism is not defined` error.  It's the `async` we're setting.  We probably want to sniff the file name for `syntax-prism` and only set `async` for that.  Let me know if I misread any of the conversation around this though.

**Comment by jwhitlock** [comment]
*TL;DR - used ``defer``, rebased on master.  @darkwing, please try again.*

To summarize the conversation:

If ``wiki.js`` detects that syntax highlighting is needed, it injects this HTML in development:

```html
<script data-manual defer src="/static/js/libs/prism/prism.js?build=1443476862"></script>
<script data-manual defer src="/static/js/prism-mdn/components/prism-json.js?build=1443476862"></script>
<script data-manual defer src="/static/js/prism-mdn/plugins/line-numbering/prism-line-numbering.js?build=1443476862"></script>
<script data-manual defer src="/static/js/libs/prism/plugins/line-highlight/prism-line-highlight.js?build=1443476862"></script>
<script data-manual defer src="/static/js/syntax-prism.js?build=1443476862"></script>
```

The ``defer`` option (which I found on [this blog article](http://davidwalsh.name/html5-async)) tells [compliant browsers](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#Browser_compatibility) (Chrome, Firefox, IE, Safari) to load scripts in the declared order.  This is needed, because later scripts (``prism-json.js``) use ``Prism``, which is declared in the first script (``prism.js``).  Otherwise, the browser tends to execute the smallest scripts first, and you get ``Prism is not defined`` errors.  On browsers that don't support ``defer``, a refresh usually works, since the browser finds the scripts in cache and executes them in order.

In production, it injects what you'd expect:
```html
<script data-manual async src="https://developer.cdn.mozilla.net/static/js/syntax-prism-min.js?build=2aab9d6"></script>
```
This is the same code as the 5-file version, except 1) they are combined into a single script, so no loading order issues, 2) the combined contents are minified, and 3) it is server from the CDN rather than Django.  Because they are combined, we use ``async`` so the page stays as responsive as possible.

**Comment by jwhitlock** [comment]
force push - *really* rebased on master

**Comment by darkwing** [comment]
Thanks for the update @jwhitlock ; `defer` is known to be flaky but since it's only for dev I think we can leave it unless we experience problems.  I'll have a fresh look on Monday and we'll roll out after merge.

**Comment by jwhitlock** [comment]
Rebased on master

**Comment by darkwing** [comment]
R+, worked well for me.  Nice work @jwhitlock !
